src/main/kotlin/cacophony/automata/DFA.kt

```diff
@@ -1,49 +1,54 @@
 package cacophony.automata
 
-interface DFA<DFAState> {
+interface DFA<DFAState, AtomType, ResultType> {
```

Comment:
in the current state, all these 3 parameters should be non nullable, right? So we could declare

```kotlin
interface DFA<DFAState : Any, AtomType : Any, ResultType : Any> {
```

Although it looks like a frustrating task to add it everywhere

You can do it or add it to list of things to do in the refactor https://github.com/mhorod/cacophony/issues/58.
We could also replace these params with interfaces, it's getting a bit too much of them in my opinion (also as part of the refactor)

Reply:
Added comment to refactor task